src/gui/render.cpp

```diff
@@ -433,10 +444,29 @@ static void RENDER_Reset( void ) {
 		}
 	}
 /* Setup the scaler variables */
-#if C_OPENGL
-	GFX_SetShader(render.shader_src);
-#endif
-	gfx_flags=GFX_SetSize(width,height,gfx_flags,gfx_scalew,gfx_scaleh,&RENDER_CallBack);
+	double par;  /* the pixel aspect ratio of the source pixel array */
+	double parn; /* normalised PAR, always > 1.0 */
+	par = (double)width / height / 4 * 3; /* MS-DOS screen is always 4:3 */
+
+	/* Enforce square pixels if aspect correction is off. Prevent absurd */
+	/* distortions by rounding to the nearest integer PAR:               */
+	if (!render.aspect) {
+		if ( par > 1.0 ) parn =       par;
+		else             parn = 1.0 / par;
+		parn = round( parn );
+		if ( par > 1.0 ) par =       parn;
+		else             par = 1.0 / parn;
```

Comment:
When `par` approaches 0, parn goes to `inf`.  If `par` is negative, parn is also negative. 

Clamp can keep `par` within sane limits; I'm not sure what the most extreme aspect ratio might be (perhaps a custom DOS flight sim that uses multiples screens wide by only one high?).  So let's say aspect can go as far as 5x in either direction.

This would solve the inf and negative issues, and also limit it to sane bounds (include `support.h` for `clamp`):

``` c++
par = clamp(par, 0.2, 5.0);
```

I also found the if/else block a bit tough to read. Wrote it down, and came up with:

``` c++
    parn = round(par > 1.0 ? par  : 1 / par);
    par  =      (par > 1.0 ? parn : 1 / parn);
```

Reply:
Although I have removed this code altogether, I will nevertheless answer your comments.

> When `par` approaches 0, parn goes to `inf`. If `par` is negative, parn is also negative.

Under no circumstances can `par` approach zero. All *PAR*s on MS-DOS were well between 1/3 and 3.
> 
> Clamp can keep `par` within sane limits; I'm not sure what the most extreme aspect ratio might be (perhaps a custom DOS flight sim for Boeing or the AirForce that uses wrap-around screens by only one screen-high?). So let's say they used 5 screens wide by 1 high (or vice-versa.. although I can't think of a vertical use-case :-) Maybe an airport with a tall "tower" of 5 screens listing flight numbers and gates?).

I don't understand. Can you show me a game with so extreme a *PAR*? I don't think it is possible. It is simply not one of the VGA or EGA modes. Am I wrong?

> I found the two if/else blocks a bit tough to read; probably because I'm not used to the formatting. Wrote it down, and came up with the follow:
> 
> ```c++
>     parn = round(par > 1.0 ? par  : 1 / par);
>     par  =      (par > 1.0 ? parn : 1 / parn);
> ```

Not bad, but I prefer to shun the redundant and unintuitive `?:` operator like the plague. If the coding standard requires that I use it, I will have to...

Reply:
> I don't understand. Can you show me a game with so extreme a PAR? I don't think it is possible. It is simply not one of the VGA or EGA modes. Am I wrong?

I was recommending using `clamp` to bound the aspect from possibly going to infinity, which the current code allows.  If `par` is ultimately calculated from user settings, then I suggest we clamp it; or at a minimum, assert it that it's within 3x either vertical or horizontal.

> Not bad, but I shun the redundant and unintuitive ?:

Yes, the two ?'s are redundant - but are fully equal to the two if statements in the original code:
```
		if ( par > 1.0 ) ... configure `parn`
		if ( par > 1.0 ) ... configure `par`
```

It was just an attempt to make it more consise and readable. Use of the `?` is not a coding standard; but it can improve readability especially when the assignment is moved outside the if/else. (`a = conditional`, instead of `conditional, a = .., else a = ...`)

If you prefer if/else's instead of `?` then please use those. No worries :-)

However, the formatting should follow K&R; which admittedly is perhaps the worst of both worlds for this snip of code?

```
if ( par > 1.0 )
	parn = par;
else
	parn = 1.0 / par;

parn = round( parn );

if ( par > 1.0 )
	par = parn;
else
	par = 1.0 / parn;
```

Reply:
*kccroft:*

> I was recommending using `clamp` to bound the aspect from possibly going to infinity, which the current code allows. If `par` is derived from user settings or external content (such as relying on the game setting a video mode), then I suggest we clamp it or at a minimum flag it `if (parn > 3.0)`.

Yes, `par` is derived from the video mode, but there are no video modes with an infinite or zero *PAR*, so it cannot go to infinity. You simply can't write such an *MS-DOS* program, which is why am loth to implement protection from this impossible situation. In yet other words, I seriously doubt you can test this boundary-checking code because it will never execute.

> If `parn` is purely derived from internal logic, then `assert` would be more applicable than `if`.

Whereas `par` is within 1/3..3, so is `parn`, so no assert is required. Of course, I may be wrong, but I need a practical reproduction of this error or at least a theoretical explanation why it is possible in at least one of *MS-DOS* graphical modes.

> ```
> parn = round(par > 1.0 ? par  : 1 / par);
> par  =      (par > 1.0 ? parn : 1 / parn);
> ```
> 
> > Not bad, but I shun the redundant and unintuitive `?:`
> 
> Fully agree: the two ?'s are redundant as they evaluate the same conditional; however the original code also has two `if`'s that evaluate the same conditional.

Of course. I say `?:` is redundant not because there are two instances of it in the code above, but because it can be replaced with the more general, universal, and intuitive `if` statement.
 
> ```
> if ( par > 1.0 ) ... configure `parn`
> else ... configure `parn`
> if ( par > 1.0 ) ... configure `par`
> else ... configure `par`
> ```
> 
> Use of the `?` is not a coding standard; but it can improve readability especially when the assignment is moved outside the conditional:
> `a = (conditional) val-1 or val-2`

I see. It is the striving of *C* towards cramping may things in one line. The treatment of assignments as expressions is part of the same tendency.
 
> Where as explicit `if / else` the assignment is moved (redundantly) inside each branch:
> `(conditional) a = val-1 else a = val-2`
> 
> If you prefer if/else's instead of `?` then please use those. No worries :-)

Yes, I do. And I will format them in multiline K&R as required.

Reply:
Thanks for explaining this @ant-222 : 
> Yes, par is derived from the video mode, but there are no video modes with an infinite or zero PAR, so it cannot go to infinity.

I thought `par` could come from the users' resolutions defined in dosbox.conf; so also fully agree it makes no sense to test or constrain this (until we have real-world examples of violations, like you said).